File: Course_2-Data_Structures/Week-3/Excercise_Challenges/1_make_heap/build_heap.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

class HeapBuilder{
private:
    vector<int> data_;
    vector<pair<int, int> > swaps_;

    void ReadData(void)
    {
        int n;
        cin >> n;
        data_.resize(n);

        for(int i = 0; i < n; ++i)
            cin >> data_[i];
    }

    int LeftChild(int index)
    {
        return 2 * index + 1;
    }

    int RightChild(int index)
    {
        return 2 * index + 2;
    }

    void ShiftDown(int index)
    {
        int n = data_.size();
        int min_index = index;
        int l = LeftChild(index);
        if(l < n && data_[l] < data_[min_index])
            min_index = l;

        int r = RightChild(index);
        if(r < n && data_[r] < data_[min_index])
            min_index = r;

        if(index != min_index){
            swaps_.push_back(make_pair(index, min_index));
            swap(data_[index], data_[min_index]);
            ShiftDown(min_index);
        }
    }

    void GenerateSwaps(void)
    {
        swaps_.clear();
        int n = data_.size();

        for(int i = n / 2; i >= 0; --i){
            ShiftDown(i);
        }
    }

    void WriteResponse(void)
    {
        cout << swaps_.size() << "\n";
        for(int i = 0; i < swaps_.size(); ++i){
            cout << swaps_[i].first << " " << swaps_[i].second << "\n";
        }
    }

public:
    void Solve(void)
    {
        ReadData();
        GenerateSwaps();
        WriteResponse();
    }
};

int main(void)
{
    ios_base::sync_with_stdio(false);
    HeapBuilder heap_builder;
    heap_builder.Solve();
  
    return 0;
}

File: Course_2-Data_Structures/Week-3/Excercise_Challenges/1_make_heap/build_heap.py
class BuildHeap:
    def __init__(self):
        self.n = 0
        self.data = []
        self.pair = []

    def read_data(self):
        self.n = int(input())
        self.data = list(map(int, input().split()))

    def left_child(self, index):
        return 2 * index + 1

    def right_child(self, index):
        return 2 * index + 2

    def shift_down(self, index):
        min_index = index

        left = self.left_child(index)
        if left < self.n and self.data[left] < self.data[min_index]:
            min_index = left

        right = self.right_child(index)
        if right < self.n and self.data[right] < self.data[min_index]:
            min_index = right
        
        if index != min_index:
            self.pair.append([index, min_index])
            self.data[index], self.data[min_index] = self.data[min_index], self.data[index]
            self.shift_down(min_index)

    def generate_swap(self):
        for i in range(self.n//2, -1, -1):
            self.shift_down(i)

    def write_pair(self):
        print(len(self.pair))
        for v in self.pair:
            print(v[0], v[1])


if __name__ == "__main__":
    obj = BuildHeap()
    obj.read_data()
    obj.generate_swap()
    obj.write_pair()

